stan/math/prim/mat/prob/ordered_logistic_log.hpp

```diff
@@ -20,37 +21,128 @@ namespace stan {
     }
 
     /**
      * @deprecated use <code>ordered_logistic_lpmf</code>
      */
-    template <typename T_lambda, typename T_cut>
-    typename boost::math::tools::promote_args<T_lambda, T_cut>::type
-    ordered_logistic_log(int y, const T_lambda& lambda,
+    template <typename T_y, typename T_loc, typename T_cut>
+    typename boost::math::tools::promote_args<T_loc, T_cut>::type
+    ordered_logistic_log(const T_y& y,
+                         const Eigen::Matrix<T_loc, Eigen::Dynamic, 1>& lambda,
                          const Eigen::Matrix<T_cut, Eigen::Dynamic, 1>& c) {
-      return ordered_logistic_lpmf<T_lambda, T_cut>(y, lambda, c);
+      return ordered_logistic_lpmf<T_loc, T_cut>(y, lambda, c);
+    }
+
+    /**
+     * Returns the (natural) log probability of the specified array 
+     * of integers given the vector of continuous locations and 
+     * array of specified cutpoints in an ordered logistic model.
+     *
+     * <p>Typically the continous location
+     * will be the dot product of a vector of regression coefficients
+     * and a vector of predictors for the outcome.
+     *
+     * @tparam propto True if calculating up to a proportion.
+     * @tparam T_y Type of y variable (should be std::vector<int>).
+     * @tparam T_loc Location type.
+     * @tparam T_cut Cut-point type.
+     * @param y Array of integers
+     * @param lambda Vector of continuous location variables.
+     * @param c array of Positive increasing vectors of cutpoints.
+     * @return Log probability of outcome given location and
+     * cutpoints.
+     * @throw std::domain_error If the outcome is not between 1 and
+     * the number of cutpoints plus 2; if the cutpoint vector is
+     * empty; if the cutpoint vector contains a non-positive,
+     * non-finite value; or if the cutpoint vector is not sorted in
+     * ascending order.
+     * @throw std::invalid_argument If y and lambda are different
+     * lengths.
+     */
+    template <bool propto, typename T_y, typename T_loc, typename T_cut>
+    typename boost::math::tools::promote_args<T_loc, T_cut>::type
+    ordered_logistic_log(const T_y& y,
+                         const Eigen::Matrix<T_loc, Eigen::Dynamic, 1>& lambda,
+                         const std::vector<Eigen::Matrix<
+                                                T_cut, Eigen::Dynamic, 1>>& c) {
+      return ordered_logistic_lpmf<propto, T_loc, T_cut>(y, lambda, c);
     }
 
+    /**
+     * @deprecated use <code>ordered_logistic_lpmf</code>
+     */
+    template <typename T_y, typename T_loc, typename T_cut>
+    typename boost::math::tools::promote_args<T_loc, T_cut>::type
+    ordered_logistic_log(const T_y& y,
+                         const Eigen::Matrix<T_loc, Eigen::Dynamic, 1>& lambda,
+                         const std::vector<Eigen::Matrix<
+                                                T_cut, Eigen::Dynamic, 1>>& c) {
+      return ordered_logistic_lpmf<T_loc, T_cut>(y, lambda, c);
+    }
```

Comment:
With the additional signatures, you really need to generalize this.

You only need two template functions.  All the work, including ruling out illegal cases, will be done by the functions you're delegating to in the body of these two functions:

```
    template <bool propto, typename T_y, typename T_loc, typename T_cut>
    typename boost::math::tools::promote_args<T_loc, T_cut>::type
    ordered_logistic_log(const T_y& y, const T_loc& lambda,
                         const T_cut& c) {
      return ordered_logistic_lpmf<propto>(y, lambda, c);
    }

    template <typename T_y, typename T_loc, typename T_cut>
    typename boost::math::tools::promote_args<T_loc, T_cut>::type
    ordered_logistic_log(const T_y& y, const T_loc& lambda,
                         const T_cut& c) {
      return ordered_logistic_lpmf(y, lambda, c);
    }
```

They should both be documented as deprecated.

I'm pretty sure you won't need the explicit template arguments in the calls in the body, but you may need to put them back if the type inference won't go through on its own.

You can generalize doc to talk about values or containers of values.

Then, you don't need to include any of the container files here (like `<vector>`).

Reply:
That template only worked for me when lambda was double, when calling it with lambda as a vector it returns the error
```
stan/lib/stan_math/stan/math/prim/mat/prob/ordered_logistic_log.hpp:15:14: error: 
      no viable conversion from returned value of type 'typename
      boost::math::tools::promote_args<double, double>::type' (aka 'double') to
      function return type 'typename
      boost::math::tools::promote_args<Matrix<double, -1, 1, 0, -1, 1>,
      vector<Matrix<double, -1, 1, 0, -1, 1>, allocator<Matrix<double, -1, 1, 0,
      -1, 1> > > >::type' (aka 'Eigen::Matrix<double, -1, 1, 0, -1, 1>')
      return ordered_logistic_lpmf<propto>(y, lambda, c);
             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
Only worked for me (for both vectorised cases) with:
```
    template <bool propto, typename T_y, typename T_loc, typename T_cut>
    typename boost::math::tools::promote_args<T_loc, T_cut>::type
    ordered_logistic_log(const T_y& y, 
                         const Eigen::Matrix<T_loc, Eigen::Dynamic, 1>& lambda,
                         const T_cut& c) {
      return ordered_logistic_lpmf<propto>(y, lambda, c);

    }

    template <typename T_y, typename T_loc, typename T_cut>
    typename boost::math::tools::promote_args<T_loc, T_cut>::type
    ordered_logistic_log(const T_y& y, 
                         const Eigen::Matrix<T_loc, Eigen::Dynamic, 1>& lambda,
                         const T_cut& c) {
      return ordered_logistic_lpmf(y, lambda, c);
    }
```
But this then doesn't work when lambda is a double, needing the separate template

Reply:
Sorry about that---needs a more general metaprogram.  You're looking for `return_type`.  Anyway, no big deal if you don't want to fix this.  It's just messy to maintain with lots of implementations.